libr/anal/fcn.c

```diff
@@ -1484,6 +1489,7 @@ static int fcn_recurse(RAnal *anal, RAnalFunction *fcn, ut64 addr, ut8 *buf, ut6
 		}
 	}
 beach:
+	check_purity (anal, fcn);
```

Comment:
Thanks for your comment. 
I chose this point to check, because to check purity of a function one should analyze recursively all the functions called by fcn either way. Since we are doing it here, there's no need to do recursion once again in some other place, in my opinion (that would cost additional time and code that is already in fcn_recurse).

Reply:
And, please correct me if I'm wrong, at that point we have all information required to analyze purity (all refs).

Reply:
That's true, the problem is.. I don't trust the radare2 analysis loop to have already analyzed all functions called by the function currently under analysis :D

If you do `r2 ./r2r/bins/elf/analysis/ls-linux64`:
```
> aa
> s main
> pd 50~call
│           0x00003afc      e87fb30000     call 0xee80
│           0x00003b0d      e80efeffff     call sym.imp.setlocale      ; char *setlocale(int category, const char *locale)
│           0x00003b20      e89bfaffff     call sym.imp.bindtextdomain ; char *bindtextdomain(char *domainname, char *dirname)
│           0x00003b2c      e84ffaffff     call sym.imp.textdomain     ; char *textdomain(char *domainname)
│           0x00003b45      e846030100     call 0x13e90
│     │││   0x00003ba5      e8f6f8ffff     call sym.imp.abort          ; void abort(void)
│      ││   0x00003baf      e84cf9ffff     call sym.imp.isatty         ; int isatty(int fd)
```

You can see there are some functions (e.g. 0xee80 and 0x13e90) that have not been analyzed, though `main` has been already analyzed as a function. This is of course a problem in radare2 and not with your idea (which is right IMHO).

Reply:
Oh, I see. I will probably try to fix this problem in the future. So, what is the point where we are 100% sure that all the called functions have been analyzed?

Reply:
Analysis is currently IMHO a bit of a mess, because it is spread across different commands (I do not know exactly why...). So there's not really a recursive disassembly algorithm in place (or is there? if there is, it's not part of `aaa`) and the only way I know to be sure all functions have been analyzed is at the end of `aaa`. And maybe you'll need yet another command just to compute the purity of the functions... @radare opinions? comments?

Reply:
imho the information of pureness should be accessed via a getter, as well as many other things in structs we hold here and there, this way we can compute the value and cache it until it's invalidated (any modification that affects this field must invalidate the pureness computation)